airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftDestination.java

```diff
@@ -13,52 +16,45 @@
- * given arguments to determine which strategy to use.
+ * The Redshift Destination offers two replication strategies. The first inserts via a typical SQL Insert statement. Although less efficient, this requires less user set up. See {@link
+ * RedshiftInsertDestination} for more detail. The second inserts via streaming the data to an S3 bucket, and Cop-ing the date into Redshift. This is more efficient, and recommended for production
+ * workloads, but does require users to set up an S3 bucket and pass in additional credentials. See {@link RedshiftStagingS3Destination} for more detail. This class inspect the given arguments to
+ * determine which strategy to use.
  */
 public class RedshiftDestination extends SwitchingDestination<RedshiftDestination.DestinationType> {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(RedshiftDestination.class);
+  private static final String METHOD = "method";
+
+  private static final Map<DestinationType, Destination> destinationMap = Map.of(
+      DestinationType.STANDARD, new RedshiftInsertDestination(),
+      DestinationType.COPY_S3, new RedshiftStagingS3Destination()
+  );
 
   enum DestinationType {
     STANDARD,
     COPY_S3
   }
 
   public RedshiftDestination() {
-    super(DestinationType.class, RedshiftDestination::getTypeFromConfig, getTypeToDestination());
+    super(DestinationType.class, RedshiftDestination::getTypeFromConfig, destinationMap);
   }
 
-  public static DestinationType getTypeFromConfig(final JsonNode config) {
+  private static DestinationType getTypeFromConfig(final JsonNode config) {
     return determineUploadMode(config);
   }
 
-  public static Map<DestinationType, Destination> getTypeToDestination() {
-    return Map.of(
-        DestinationType.STANDARD, new RedshiftInsertDestination(),
-        DestinationType.COPY_S3, new RedshiftStagingS3Destination());
-  }
-
   public static DestinationType determineUploadMode(final JsonNode config) {
-    final var bucketNode = config.get("s3_bucket_name");
-    final var regionNode = config.get("s3_bucket_region");
-    final var accessKeyIdNode = config.get("access_key_id");
-    final var secretAccessKeyNode = config.get("secret_access_key");
 
-    if (isNullOrEmpty(bucketNode) && isNullOrEmpty(regionNode) && isNullOrEmpty(accessKeyIdNode)
-        && isNullOrEmpty(secretAccessKeyNode)) {
+    final JsonNode jsonNode = doesJsonNodeContainUploadingMethod(config);
+
+    if (validateIfAllRequiredS3fieldsAreNullOrEmpty(jsonNode)) {
       LOGGER.warn("The \"standard\" upload mode is not performant, and is not recommended for production. " +
           "Please use the Amazon S3 upload mode if you are syncing a large amount of data.");
       return DestinationType.STANDARD;
     }
 
-    if (isNullOrEmpty(bucketNode) && isNullOrEmpty(regionNode) && isNullOrEmpty(accessKeyIdNode)
-        && isNullOrEmpty(secretAccessKeyNode)) {
+    if (validateIfAllRequiredS3fieldsAreNullOrEmpty(jsonNode)) {
```

Comment:
hm, I think this might have been bugged previously? It seems weird that this condition is identical to the previous one. Maybe it was supposed to be using `||` instead of `&&`?

Reply:
Sorry, the statement by itself is correct just a little bit confused with the name. Fixed please take a look.

__
[![sema-logo](https://app.semasoftware.com/img/sema-tray-logo.gif)](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix

Reply:
I think I'm confused about when this condition is ever true: the if-statement right above has the exact same condition, so wouldn't we always hit the `return Destination.STANDARD`?

Reply:
@edgao if each of the next parameters:
```java
   return isNullOrEmpty(jsonNode.get("s3_bucket_name"))
        && isNullOrEmpty(jsonNode.get("s3_bucket_region"))
        && isNullOrEmpty(jsonNode.get("access_key_id"))
        && isNullOrEmpty(jsonNode.get("secret_access_key"));
```
**won't be null or empty**, the method will return `DestinationType.COPY_S3`

However, I strongly agree with you we don't need to check this parameter on BE, as if any of them is null or empty we will strike `Destination.STANDARD` anyway.